cmd/ssl-selfsign-handshake-check/main.go

```diff
@@ -42,23 +30,79 @@ func init() {
 }
 
 func main() {
-	err := runSelfsignHandshake()
+	// create context
+	checkTimeLimit := time.Minute * 1
+	ctx, _ := context.WithTimeout(context.Background(), checkTimeLimit)
+
+	// create Kubernetes client
+	kubernetesClient, err := kubeClient.Create("")
+	if err != nil {
+		log.Errorln("Error creating kubeClient with error" + err.Error())
+	}
+
+	// hits kuberhealthy endpoint to see if node is ready
+	err = nodeCheck.WaitForKuberhealthy(ctx)
+	if err != nil {
+		log.Errorln("Error waiting for kuberhealthy endpoint to be contactable by checker pod with error:" + err.Error())
+	}
+
+	// fetches kube proxy to see if it is ready
+	err = nodeCheck.WaitForKubeProxy(ctx, kubernetesClient, "kuberhealthy", "kube-system")
+	if err != nil {
+		log.Errorln("Error waiting for kube proxy to be ready and running on the node with error:" + err.Error())
+	}
+	fileFound := whichCheck()
+	err = runCheck(fileFound)
 	if err != nil {
 		reportErr := reportKHFailure(err.Error())
 		if reportErr != nil {
 			log.Error(reportErr)
 		}
 		os.Exit(1)
 	}
-	reportErr := reportKHSuccess()
-	if reportErr != nil {
-		log.Error(reportErr)
+
+	if err == nil {
+		reportErr := reportKHSuccess()
+		if reportErr != nil {
+			log.Error(reportErr)
+		}
+		os.Exit(1)
+	}
+}
+
+// whichCheck determines if a cert has been provided, or if it needs to be retrieved from the host and returns a bool
+func whichCheck() bool {
+	var fromFile bool
+	if _, err := os.Stat("/etc/ssl/selfsign/certificate.crt"); err == nil {
+		fromFile = true
 	}
+	return fromFile
 }
 
-// run the SSL handshake check for the specified host and port number from ssl_util package
-func runSelfsignHandshake() error {
-	err := ssl_util.SelfsignHandshake(domainName, portNum)
+// runCheck takes in a bool from the whichCheck func and calls the one of two handshake check functions
+func runCheck(fileFound bool) error {
+	var err error
+
+	// if the user provided pem file is present, import it and run the handshake check
+	if fileFound == true {
+		err := ssl_util.HandshakeFromFile(domainName, portNum)
+		return err
+	}
+
+	// if the pem file has not been provided, pull the cert from the host, import it, and for the handshake check
+	if fileFound == false {
+		selfCert, err := ssl_util.CertificatePuller(domainName, portNum)
+		if err != nil {
+			log.Warn("Error pulling certificate from host: ", err)
+		}
+
+		err = ssl_util.HandshakeFromHost(domainName, portNum, selfCert)
+
+		if err != nil {
+			log.Warn("Error performing handshake: ", err)
+		}
+		return err
+	}
 	return err
```

Comment:
if you don't instantiate the `var err error` in the beginning you can just `return nil`